planner/core/preprocess.go

```diff
@@ -1759,3 +1763,18 @@ func (p *preprocessor) initTxnContextProviderIfNecessary(node ast.Node) {
 		InfoSchema: p.ensureInfoSchema(),
 	})
 }
+
+func (p *preprocessor) tryAutoConvert(colDef *ast.ColumnDef) bool {
+	sessVars := p.ctx.GetSessionVars()
+	if !sessVars.SQLMode.HasStrictMode() && colDef.Tp.Tp == mysql.TypeVarchar &&
+		colDef.Tp.Charset != "" {
```

Comment:
Why do we need `colDef.Tp.Charset != ""` here?  Could we add a test?

Reply:
It may not be needed and can be removed. See also https://github.com/pingcap/tidb/pull/31974#discussion_r814436827